### PR TITLE
Solve handling of wrong length status codes with SM card bug and Add ECC support in Feitian ePass2003

### LINKED ISSUE #1073: ePass2003 EC not supported
It seems that OpenSC (**0.16.0** Arch Linux build) don't support elliptic curves on Feitian Technologies, Inc. ePass2003 USB token (**096e:0807**).

    ~> opensc-tool --list-algorithms
    Using reader with a card: Feitian ePass2003 00 00
    Algorithm: rsa
    Key length: 512
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 768
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 1024
    Flags: onboard key generation padding ( none ) hashes ( )

    Algorithm: rsa
    Key length: 2048
    Flags: onboard key generation padding ( none ) hashes ( )

    ~> pkcs11-tool -k --key-type EC:prime256v1
    Using slot 0 with a present token (0x0)
    error: Generate EC key mechanism not supported

    ~> pkcs15-init --generate-key EC:prime256v1 --key-usage sign --auth-id 1
    Using reader with a card: Feitian ePass2003 00 00
    Failed to generate key: Not supported

Bay the way: why is epass2003.profile described as `pkcs15 profile for entersafe`

### LINKED ISSUE #1115: feitian epass2003 is not working with opensc 0.17.0 on Fedora 26
```
opensc-tool --list-readers                                                                                                                               
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             FT ePass2003Auto 00 00
```

pkcs15-init --erase-card -vvv  **(this works fine)**
pkcs15-init --create-pkcs15 --profile pkcs15+onepin --label "sandys" -vvv **(this works fine)**
pkcs15-init --auth-id 1 --generate-key rsa/2048 --key-usage sign,decrypt -vvv **(this fails)**

> Using reader with a card: FT ePass2003Auto 00 00
> PKCS#15 binding failed: Unsupported card
> 

the detailed logs have been attached. grep for "sss>" to see the different commands that i have run separately.
[debug.log.zip](https://github.com/OpenSC/OpenSC/files/1191090/debug.log.zip)

> 
> uname -a                                                                                                                                                 
> Linux menzoberranzan 4.11.11-300.fc26.x86_64 #1 SMP Mon Jul 17 16:32:11 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
> 
> LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
> Distributor ID:	Fedora
> Description:	Fedora release 26 (Twenty Six)
> Release:	26
> Codename:	TwentySix

> 
> Installed Packages
> Name         : opensc
> Version      : 0.17.0
> Release      : 1.fc26
> Arch         : x86_64
> Size         : 3.3 M
> Source       : opensc-0.17.0-1.fc26.src.rpm

> Name         : pcsc-tools
> Version      : 1.4.25
> Release      : 2.fc26
> Arch         : x86_64
> Size         : 836 k
> Source       : pcsc-tools-1.4.25-2.fc26.src.rpm
> 
> Name         : pcsc-lite-ccid
> Version      : 1.4.27
> Release      : 1.fc26
> Arch         : x86_64
> Size         : 1.4 M

### PR DESCRIPTION
**Pull request by FeitianSmartcardReader**
Fixes #1073 
Fixes #1115 

##### Checklist

- [x] Documentation is added or updated
- [x] New files have a LGPL 2.1 license statement
- [x] Tested with the following card: 
	- [x] tested PKCS#11
	- [ ] tested Windows Minidriver
	- [x] tested macOS Tokend

### CODE REVIEW

File: src/libopensc/card-epass2003.c [hunk 1]
```diff
@@ -1192,6 +1202,11 @@ epass2003_init(struct sc_card *card)
 	_sc_card_add_rsa_alg(card, 1024, flags, 0);
 	_sc_card_add_rsa_alg(card, 2048, flags, 0);
 
+	//set EC Alg Flags
+	flags = SC_ALGORITHM_ONBOARD_KEY_GEN|SC_ALGORITHM_ECDSA_HASH_SHA1|SC_ALGORITHM_ECDSA_HASH_SHA256|SC_ALGORITHM_ECDSA_HASH_NONE|SC_ALGORITHM_ECDSA_RAW;
+	//ext_flags = SC_ALGORITHM_EXT_EC_NAMEDCURVE | SC_ALGORITHM_EXT_EC_UNCOMPRESES;
```
Comment (majkrzak) [inline src/libopensc/card-epass2003.c]: Initialize `ext_flags` to zero or to this commented flags.

File: src/libopensc/card-epass2003.c [hunk 2]
```diff
@@ -170,6 +172,7 @@ static const struct sc_card_error epass2003_errors[] = {
 static int epass2003_transmit_apdu(struct sc_card *card, struct sc_apdu *apdu);
 static int epass2003_select_file(struct sc_card *card, const sc_path_t * in_path, sc_file_t ** file_out);
 int epass2003_refresh(struct sc_card *card);
+static int hash_data(unsigned char *data, size_t datalen, unsigned char *hash, unsigned int mechanismType);
```
Comment (majkrzak) [inline src/libopensc/card-epass2003.c]: Add a `const` specifier to `data` param.

File: src/libopensc/card-epass2003.c [hunk 3]
```diff
@@ -2349,6 +2483,7 @@ epass2003_card_ctl(struct sc_card *card, unsigned long cmd, void *ptr)
 {
 	LOG_FUNC_CALLED(card->ctx);
 
+	sc_log(card->ctx, "cmd is %0x", cmd);
```
Comment (majkrzak) [inline src/libopensc/card-epass2003.c]: Replace format specifier with `%0lx`.

File: src/libopensc/card-epass2003.c [hunk 4]
```diff
@@ -2129,19 +2235,36 @@ internal_write_rsa_key(struct sc_card *card, unsigned short fid, struct sc_pkcs1
 
 
 static int
-hash_data(unsigned char *data, size_t datalen, unsigned char *hash)
+hash_data(unsigned char *data, size_t datalen, unsigned char *hash, unsigned int mechanismType)
```
Comment (majkrzak) [inline src/libopensc/card-epass2003.c]: Add a `const` specifier to `data` param.

**Comment by FeitianSmartcardReader** [comment]
To make it clearly, close this PR, will open another PR, thanks

**Pull request closed without merging**